infrastructure/src/main/java/org/corfudb/infrastructure/redundancy/RedundancyCalculator.java

```diff
@@ -0,0 +1,339 @@
+package org.corfudb.infrastructure.redundancy;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Sets;
+import lombok.AllArgsConstructor;
+import lombok.EqualsAndHashCode;
+import lombok.Getter;
+import lombok.NonNull;
+import lombok.ToString;
+import org.corfudb.infrastructure.log.statetransfer.StateTransferManager.CurrentTransferSegment;
+import org.corfudb.infrastructure.log.statetransfer.StateTransferManager.SegmentState;
+import org.corfudb.runtime.view.Layout;
+import org.corfudb.runtime.view.Layout.LayoutSegment;
+import org.corfudb.runtime.view.Layout.LayoutStripe;
+import org.corfudb.runtime.view.LayoutBuilder;
+
+import java.util.List;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+import static org.corfudb.infrastructure.log.statetransfer.StateTransferManager.CurrentTransferSegmentStatus;
+import static org.corfudb.infrastructure.log.statetransfer.StateTransferManager.SegmentState.FAILED;
+import static org.corfudb.infrastructure.log.statetransfer.StateTransferManager.SegmentState.RESTORED;
+import static org.corfudb.infrastructure.log.statetransfer.StateTransferManager.SegmentState.TRANSFERRED;
+import static org.corfudb.infrastructure.redundancy.RedundancyCalculator.SegmentAge.CURRENT_EPOCH;
+import static org.corfudb.infrastructure.redundancy.RedundancyCalculator.SegmentAge.PREVIOUS_EPOCH;
+
+/**
+ * A class used to compute the transfer segments, as well as update the layout
+ * after the redundancy restoration.
+ */
+@AllArgsConstructor
+public class RedundancyCalculator {
+
+    /**
+     * A status that tells whether a transfer segment belongs to the current or a previous epoch.
+     */
+    public enum SegmentAge {
+        PREVIOUS_EPOCH,
+        CURRENT_EPOCH
+    }
+
+    /**
+     * A transfer segment that also contains an information about its age.
+     */
+    @AllArgsConstructor
+    @Getter
+    @ToString
+    @EqualsAndHashCode
+    public static class AgedSegment implements Comparable<AgedSegment> {
+        private final CurrentTransferSegment segment;
+        private final SegmentAge age;
+
+        @Override
+        public int compareTo(AgedSegment other) {
+            return (int) (this.getSegment().getStartAddress() -
```

Comment:
let's make `CurrentTransferSegment ` comparable. Then you can just compare them, like 
```
segment.compareTo(other.segment)
```

or at least this can be replaced by
```
Long.compare(segment.getStartAddress(), other.segment.getStartAddress())
```